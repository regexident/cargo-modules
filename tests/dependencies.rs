#[macro_use]
mod util;

mod help {
    test_cmd!(
        args: "dependencies \
                --help",
        success: true,
        color_mode: ColorMode::Plain,
        project: smoke
    );
}

mod colors {
    mod plain {
        test_cmd!(
            args: "dependencies",
            success: true,
            color_mode: ColorMode::Plain,
            project: smoke
        );
    }

    mod ansi {
        test_cmd!(
            args: "dependencies",
            success: true,
            color_mode: ColorMode::Ansi,
            project: smoke
        );
    }

    mod truecolor {
        test_cmd!(
            args: "dependencies",
            success: true,
            color_mode: ColorMode::TrueColor,
            project: smoke
        );
    }
}

mod default {
    mod pass {
        test_cmds!(
            args: "dependencies",
            success: true,
            color_mode: ColorMode::Plain,
            projects: [
                package_bin_target,
                package_lib_target,
                virtual_workspace_single_package_lib_target,
                virtual_workspace_single_package_bin_target,
                workspace_single_package_lib_target,
                workspace_single_package_bin_target,
            ]
        );
    }

    mod fail {
        test_cmds!(
            args: "dependencies",
            success: false,
            color_mode: ColorMode::Plain,
            projects: [
                package_multi_target,
                virtual_workspace_single_package_multi_target,
                workspace_single_package_multi_target,
            ]
        );
    }
}

mod lib {
    mod pass {
        test_cmds!(
            args: "dependencies \
                    --lib",
            success: true,
            color_mode: ColorMode::Plain,
            projects: [
                package_lib_target,
                package_multi_target,
                virtual_workspace_single_package_lib_target,
                virtual_workspace_single_package_multi_target,
                workspace_single_package_lib_target,
                workspace_single_package_multi_target,
            ]
        );
    }

    mod fail {
        test_cmds!(
            args: "dependencies \
                    --lib", // does not exist
            success: false,
            color_mode: ColorMode::Plain,
            projects: [
                package_bin_target,
                virtual_workspace_single_package_bin_target,
                workspace_single_package_bin_target,
            ]
        );
    }
}

mod bin {
    mod pass {
        test_cmds!(
            args: "dependencies \
                    --bin package_bin_target",
            success: true,
            color_mode: ColorMode::Plain,
            projects: [
                package_bin_target,
                virtual_workspace_single_package_bin_target,
                workspace_single_package_bin_target,
            ]
        );

        test_cmds!(
            args: "dependencies \
                    --bin package_multi_target",
            success: true,
            color_mode: ColorMode::Plain,
            projects: [
                package_multi_target,
                virtual_workspace_single_package_multi_target,
                workspace_single_package_multi_target,
            ]
        );
    }

    mod fail {
        test_cmds!(
            args: "dependencies \
                    --bin foobar", // does not exist
            success: false,
            color_mode: ColorMode::Plain,
            projects: [
                package_lib_target,
                package_multi_target,
                virtual_workspace_multi_package,
                virtual_workspace_single_package_lib_target,
                virtual_workspace_single_package_multi_target,
                workspace_multi_package,
                workspace_single_package_lib_target,
                workspace_single_package_multi_target,
            ]
        );
    }
}

mod package {
    mod pass {
        test_cmds!(
            args: "dependencies \
                    --package package_lib_target",
            success: true,
            color_mode: ColorMode::Plain,
            projects: [
                package_lib_target,
                virtual_workspace_single_package_lib_target,
                workspace_single_package_lib_target,
            ]
        );

        test_cmds!(
            args: "dependencies \
                    --package package_bin_target",
            success: true,
            color_mode: ColorMode::Plain,
            projects: [
                package_bin_target,
                virtual_workspace_single_package_bin_target,
                workspace_single_package_bin_target,
            ]
        );
    }

    mod fail {
        test_cmds!(
            args: "dependencies \
                    --package foobar",
            success: false,
            color_mode: ColorMode::Plain,
            projects: [
                package_bin_target,
                package_lib_target,
                package_multi_target,
                virtual_workspace_multi_package,
                virtual_workspace_single_package_bin_target,
                virtual_workspace_single_package_lib_target,
                virtual_workspace_single_package_multi_target,
                workspace_multi_package,
                workspace_single_package_bin_target,
                workspace_single_package_lib_target,
                workspace_single_package_multi_target,
            ]
        );
    }
}

mod package_lib {
    mod pass {
        test_cmds!(
            args: "dependencies \
                    --package package_lib_target \
                    --lib",
            success: true,
            color_mode: ColorMode::Plain,
            projects: [
                package_lib_target,
                workspace_single_package_lib_target,
                virtual_workspace_single_package_lib_target,
            ]
        );

        test_cmds!(
            args: "dependencies \
                    --package package_multi_target \
                    --lib",
            success: true,
            color_mode: ColorMode::Plain,
            projects: [
                package_multi_target,
                workspace_single_package_multi_target,
                virtual_workspace_single_package_multi_target,
            ]
        );
    }

    mod fail {
        test_cmds!(
            args: "dependencies \
                    --package package_bin_target \
                    --lib", // does not exist
            success: false,
            color_mode: ColorMode::Plain,
            projects: [
                package_bin_target,
                workspace_single_package_bin_target,
                virtual_workspace_single_package_bin_target,
            ]
        );
    }
}

mod package_bin {
    mod pass {
        test_cmds!(
            args: "dependencies \
                    --package package_bin_target \
                    --bin package_bin_target",
            success: true,
            color_mode: ColorMode::Plain,
            projects: [
                package_bin_target,
                workspace_single_package_bin_target,
                virtual_workspace_single_package_bin_target,
            ]
        );

        test_cmds!(
            args: "dependencies \
                    --package package_multi_target \
                    --bin package_multi_target",
            success: true,
            color_mode: ColorMode::Plain,
            projects: [
                package_multi_target,
                workspace_single_package_multi_target,
                virtual_workspace_single_package_multi_target,
            ]
        );
    }

    mod fail {
        test_cmds!(
            args: "dependencies \
                    --package workspace-package \
                    --bin foobar", // does not exist
            success: false,
            color_mode: ColorMode::Plain,
            projects: [
                package_bin_target,
                package_lib_target,
                package_multi_target,
                virtual_workspace_multi_package,
                virtual_workspace_single_package_bin_target,
                virtual_workspace_single_package_lib_target,
                virtual_workspace_single_package_multi_target,
                workspace_multi_package,
                workspace_single_package_bin_target,
                workspace_single_package_lib_target,
                workspace_single_package_multi_target,
            ]
        );
    }
}

mod cfg_test {
    mod without_tests {
        test_cmd!(
            args: "dependencies",
            success: true,
            color_mode: ColorMode::Plain,
            project: smoke
        );
    }

    mod with_tests {
        test_cmd!(
            args: "dependencies \
                    --cfg-test",
            success: true,
            color_mode: ColorMode::Plain,
            project: smoke
        );
    }
}

mod selection {
    mod no_externs {
        test_cmd!(
            args: "dependencies \
                    --no-externs",
            success: true,
            color_mode: ColorMode::Plain,
            project: smoke
        );
    }

    mod no_fns {
        test_cmd!(
            args: "dependencies \
                    --no-fns",
            success: true,
            color_mode: ColorMode::Plain,
            project: smoke
        );
    }

    mod no_modules {
        test_cmd!(
            args: "dependencies \
                    --no-modules",
            success: true,
            color_mode: ColorMode::Plain,
            project: smoke
        );
    }

    mod no_owns {
        test_cmd!(
            args: "dependencies \
                    --no-owns",
            success: true,
            color_mode: ColorMode::Plain,
            project: smoke
        );
    }

    mod no_traits {
        test_cmd!(
            args: "dependencies \
                    --no-traits",
            success: true,
            color_mode: ColorMode::Plain,
            project: smoke
        );
    }

    mod no_types {
        test_cmd!(
            args: "dependencies \
                    --no-types",
            success: true,
            color_mode: ColorMode::Plain,
            project: smoke
        );
    }

    mod no_uses {
        test_cmd!(
            args: "dependencies \
                    --no-uses",
            success: true,
            color_mode: ColorMode::Plain,
            project: smoke
        );
    }
}

mod focus_on {
    mod simple_path {
        test_cmd!(
            args: "dependencies \
                    --focus-on \"crate::visibility::dummy\"",
            success: true,
            color_mode: ColorMode::Plain,
            project: smoke
        );
    }

    mod glob_path {
        test_cmd!(
            args: "dependencies \
                    --focus-on \"crate::visibility::*\"",
            success: true,
            color_mode: ColorMode::Plain,
            project: smoke
        );
    }

    mod self_path {
        test_cmd!(
            args: "dependencies \
                    --focus-on \"crate::visibility::dummy::{self}\"",
            success: true,
            color_mode: ColorMode::Plain,
            project: smoke
        );
    }

    mod use_tree {
        test_cmd!(
            args: "dependencies \
                    --focus-on \"crate::visibility::{dummy, hierarchy}\"",
            success: true,
            color_mode: ColorMode::Plain,
            project: smoke
        );
    }

    mod nonexistent_path {
        test_cmd!(
            args: "dependencies \
                    --focus-on \"nonexistent\"",
            success: false,
            color_mode: ColorMode::Plain,
            project: smoke
        );
    }
}

mod max_depth {
    mod depth_0 {
        test_cmd!(
            args: "dependencies \
                    --max-depth 0",
            success: true,
            color_mode: ColorMode::Plain,
            project: smoke
        );
    }

    mod depth_1 {
        test_cmd!(
            args: "dependencies \
                    --max-depth 1",
            success: true,
            color_mode: ColorMode::Plain,
            project: smoke
        );
    }

    mod depth_2 {
        test_cmd!(
            args: "dependencies \
                    --max-depth 2",
            success: true,
            color_mode: ColorMode::Plain,
            project: smoke
        );
    }
}

mod fields {
    test_cmds!(
        args: "dependencies",
        success: true,
        color_mode: ColorMode::Plain,
        projects: [
            enum_fields,
            struct_fields,
            tuple_fields,
            union_fields,
        ]
    );
}

mod functions {
    test_cmds!(
        args: "dependencies",
        success: true,
        color_mode: ColorMode::Plain,
        projects: [
            function_inputs,
            function_outputs,
            function_body,
        ]
    );
}

mod github {
    mod issue_79 {
        test_cmd!(
            args: "dependencies \
                    --no-externs \
                    --no-fns \
                    --no-traits \
                    --no-types",
            success: true,
            color_mode: ColorMode::Plain,
            project: github_issue_79
        );
    }

    mod issue_80 {
        mod tests {
            test_cmd!(
                args: "dependencies \
                        --cfg-test \
                        --no-externs \
                        --no-fns \
                        --no-traits",
                success: true,
                color_mode: ColorMode::Plain,
                project: github_issue_80
            );
        }

        mod without_tests {
            test_cmd!(
                args: "dependencies \
                        --no-externs \
                        --no-fns \
                        --no-traits",
                success: true,
                color_mode: ColorMode::Plain,
                project: github_issue_80
            );
        }
    }

    mod issue_102 {
        test_cmd!(
            args: "dependencies \
                    --no-externs \
                    --no-fns \
                    --no-traits \
                    --no-types",
            success: true,
            color_mode: ColorMode::Plain,
            project: github_issue_102
        );
    }

    mod issue_172 {
        test_cmd!(
            args: "dependencies \
                    --no-externs \
                    --no-fns \
                    --layout dot",
            success: true,
            color_mode: ColorMode::Plain,
            project: github_issue_172
        );
    }

    mod issue_362 {
        test_cmd!(
            args: "dependencies \
                    --features 'opt-in'",
            success: true,
            color_mode: ColorMode::Plain,
            project: github_issue_362
        );
    }
}
