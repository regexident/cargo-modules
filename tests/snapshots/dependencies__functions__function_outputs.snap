---
source: tests/dependencies.rs
expression: output
---
CMD:
dependencies

STDERR:

STDOUT:
digraph {

    graph [
        label="function_outputs",
        labelloc=t,

        pad=0.4,

        // Consider rendering the graph using a different layout algorithm, such as:
        // [dot, neato, twopi, circo, fdp, sfdp]
        layout=neato,
        overlap=false,
        splines="line",
        rankdir=LR,

        fontname="Helvetica", 
        fontsize="36",
    ];

    node [
        fontname="monospace",
        fontsize="10",
        shape="record",
        style="filled",
    ];

    edge [
        fontname="monospace",
        fontsize="10",
    ];

    "bool" [label="external builtin|bool", fillcolor="#81c169"]; // "builtin" node
    "char" [label="external builtin|char", fillcolor="#81c169"]; // "builtin" node
    "f64" [label="external builtin|f64", fillcolor="#81c169"]; // "builtin" node
    "function_outputs" [label="crate|function_outputs", fillcolor="#5397c8"]; // "crate" node
    "function_outputs::Dummy" [label="pub(crate) struct|function_outputs::Dummy", fillcolor="#f8c04c"]; // "struct" node
    "function_outputs::Dummy::method_output" [label="pub(crate) fn|function_outputs::Dummy::method_output", fillcolor="#f8c04c"]; // "fn" node
    "function_outputs::GenericTargetEnum" [label="pub(crate) enum|function_outputs::GenericTargetEnum", fillcolor="#f8c04c"]; // "enum" node
    "function_outputs::GenericTargetEnum::clone" [label="pub(crate) fn|function_outputs::GenericTargetEnum::clone", fillcolor="#f8c04c"]; // "fn" node
    "function_outputs::GenericTargetStruct" [label="pub(crate) struct|function_outputs::GenericTargetStruct", fillcolor="#f8c04c"]; // "struct" node
    "function_outputs::GenericTargetStruct::clone" [label="pub(crate) fn|function_outputs::GenericTargetStruct::clone", fillcolor="#f8c04c"]; // "fn" node
    "function_outputs::GenericTargetUnion" [label="pub(crate) union|function_outputs::GenericTargetUnion", fillcolor="#f8c04c"]; // "union" node
    "function_outputs::GenericTargetUnion::clone" [label="pub(crate) fn|function_outputs::GenericTargetUnion::clone", fillcolor="#f8c04c"]; // "fn" node
    "function_outputs::GenericTypeAlias" [label="pub(crate) type|function_outputs::GenericTypeAlias", fillcolor="#f8c04c"]; // "type" node
    "function_outputs::TargetEnum" [label="pub(crate) enum|function_outputs::TargetEnum", fillcolor="#f8c04c"]; // "enum" node
    "function_outputs::TargetEnum::clone" [label="pub(crate) fn|function_outputs::TargetEnum::clone", fillcolor="#f8c04c"]; // "fn" node
    "function_outputs::TargetStruct" [label="pub(crate) struct|function_outputs::TargetStruct", fillcolor="#f8c04c"]; // "struct" node
    "function_outputs::TargetStruct::clone" [label="pub(crate) fn|function_outputs::TargetStruct::clone", fillcolor="#f8c04c"]; // "fn" node
    "function_outputs::TargetTrait" [label="pub(crate) trait|function_outputs::TargetTrait", fillcolor="#f8c04c"]; // "trait" node
    "function_outputs::TargetUnion" [label="pub(crate) union|function_outputs::TargetUnion", fillcolor="#f8c04c"]; // "union" node
    "function_outputs::TargetUnion::clone" [label="pub(crate) fn|function_outputs::TargetUnion::clone", fillcolor="#f8c04c"]; // "fn" node
    "function_outputs::TypeAlias" [label="pub(crate) type|function_outputs::TypeAlias", fillcolor="#f8c04c"]; // "type" node
    "function_outputs::function" [label="pub(crate) fn|function_outputs::function", fillcolor="#f8c04c"]; // "fn" node
    "function_outputs::function_output" [label="pub(crate) fn|function_outputs::function_output", fillcolor="#f8c04c"]; // "fn" node
    "function_outputs::generic_function" [label="pub(crate) fn|function_outputs::generic_function", fillcolor="#f8c04c"]; // "fn" node
    "isize" [label="external builtin|isize", fillcolor="#81c169"]; // "builtin" node
    "str" [label="external builtin|str", fillcolor="#81c169"]; // "builtin" node
    "usize" [label="external builtin|usize", fillcolor="#81c169"]; // "builtin" node

    "function_outputs" -> "function_outputs::Dummy" [label="owns", color="#000000", style="solid"] [constraint=true]; // "owns" edge
    "function_outputs" -> "function_outputs::GenericTargetEnum" [label="owns", color="#000000", style="solid"] [constraint=true]; // "owns" edge
    "function_outputs" -> "function_outputs::GenericTargetStruct" [label="owns", color="#000000", style="solid"] [constraint=true]; // "owns" edge
    "function_outputs" -> "function_outputs::GenericTargetUnion" [label="owns", color="#000000", style="solid"] [constraint=true]; // "owns" edge
    "function_outputs" -> "function_outputs::GenericTypeAlias" [label="owns", color="#000000", style="solid"] [constraint=true]; // "owns" edge
    "function_outputs" -> "function_outputs::TargetEnum" [label="owns", color="#000000", style="solid"] [constraint=true]; // "owns" edge
    "function_outputs" -> "function_outputs::TargetStruct" [label="owns", color="#000000", style="solid"] [constraint=true]; // "owns" edge
    "function_outputs" -> "function_outputs::TargetTrait" [label="owns", color="#000000", style="solid"] [constraint=true]; // "owns" edge
    "function_outputs" -> "function_outputs::TargetUnion" [label="owns", color="#000000", style="solid"] [constraint=true]; // "owns" edge
    "function_outputs" -> "function_outputs::TypeAlias" [label="owns", color="#000000", style="solid"] [constraint=true]; // "owns" edge
    "function_outputs" -> "function_outputs::function" [label="owns", color="#000000", style="solid"] [constraint=true]; // "owns" edge
    "function_outputs" -> "function_outputs::function_output" [label="owns", color="#000000", style="solid"] [constraint=true]; // "owns" edge
    "function_outputs" -> "function_outputs::generic_function" [label="owns", color="#000000", style="solid"] [constraint=true]; // "owns" edge
    "function_outputs::Dummy" -> "function_outputs::Dummy::method_output" [label="owns", color="#000000", style="solid"] [constraint=true]; // "owns" edge
    "function_outputs::Dummy::method_output" -> "bool" [label="uses", color="#7f7f7f", style="dashed"] [constraint=false]; // "uses" edge
    "function_outputs::Dummy::method_output" -> "char" [label="uses", color="#7f7f7f", style="dashed"] [constraint=false]; // "uses" edge
    "function_outputs::Dummy::method_output" -> "f64" [label="uses", color="#7f7f7f", style="dashed"] [constraint=false]; // "uses" edge
    "function_outputs::Dummy::method_output" -> "function_outputs::GenericTargetEnum" [label="uses", color="#7f7f7f", style="dashed"] [constraint=false]; // "uses" edge
    "function_outputs::Dummy::method_output" -> "function_outputs::GenericTargetStruct" [label="uses", color="#7f7f7f", style="dashed"] [constraint=false]; // "uses" edge
    "function_outputs::Dummy::method_output" -> "function_outputs::GenericTargetUnion" [label="uses", color="#7f7f7f", style="dashed"] [constraint=false]; // "uses" edge
    "function_outputs::Dummy::method_output" -> "function_outputs::TargetEnum" [label="uses", color="#7f7f7f", style="dashed"] [constraint=false]; // "uses" edge
    "function_outputs::Dummy::method_output" -> "function_outputs::TargetStruct" [label="uses", color="#7f7f7f", style="dashed"] [constraint=false]; // "uses" edge
    "function_outputs::Dummy::method_output" -> "function_outputs::TargetTrait" [label="uses", color="#7f7f7f", style="dashed"] [constraint=false]; // "uses" edge
    "function_outputs::Dummy::method_output" -> "function_outputs::TargetUnion" [label="uses", color="#7f7f7f", style="dashed"] [constraint=false]; // "uses" edge
    "function_outputs::Dummy::method_output" -> "isize" [label="uses", color="#7f7f7f", style="dashed"] [constraint=false]; // "uses" edge
    "function_outputs::Dummy::method_output" -> "str" [label="uses", color="#7f7f7f", style="dashed"] [constraint=false]; // "uses" edge
    "function_outputs::Dummy::method_output" -> "usize" [label="uses", color="#7f7f7f", style="dashed"] [constraint=false]; // "uses" edge
    "function_outputs::GenericTargetEnum" -> "function_outputs::GenericTargetEnum::clone" [label="owns", color="#000000", style="solid"] [constraint=true]; // "owns" edge
    "function_outputs::GenericTargetEnum::clone" -> "function_outputs::GenericTargetEnum" [label="uses", color="#7f7f7f", style="dashed"] [constraint=false]; // "uses" edge
    "function_outputs::GenericTargetStruct" -> "function_outputs::GenericTargetStruct::clone" [label="owns", color="#000000", style="solid"] [constraint=true]; // "owns" edge
    "function_outputs::GenericTargetStruct::clone" -> "function_outputs::GenericTargetStruct" [label="uses", color="#7f7f7f", style="dashed"] [constraint=false]; // "uses" edge
    "function_outputs::GenericTargetUnion" -> "function_outputs::GenericTargetUnion::clone" [label="owns", color="#000000", style="solid"] [constraint=true]; // "owns" edge
    "function_outputs::GenericTargetUnion::clone" -> "function_outputs::GenericTargetUnion" [label="uses", color="#7f7f7f", style="dashed"] [constraint=false]; // "uses" edge
    "function_outputs::TargetEnum" -> "function_outputs::TargetEnum::clone" [label="owns", color="#000000", style="solid"] [constraint=true]; // "owns" edge
    "function_outputs::TargetEnum::clone" -> "function_outputs::TargetEnum" [label="uses", color="#7f7f7f", style="dashed"] [constraint=false]; // "uses" edge
    "function_outputs::TargetStruct" -> "function_outputs::TargetStruct::clone" [label="owns", color="#000000", style="solid"] [constraint=true]; // "owns" edge
    "function_outputs::TargetStruct::clone" -> "function_outputs::TargetStruct" [label="uses", color="#7f7f7f", style="dashed"] [constraint=false]; // "uses" edge
    "function_outputs::TargetUnion" -> "function_outputs::TargetUnion::clone" [label="owns", color="#000000", style="solid"] [constraint=true]; // "owns" edge
    "function_outputs::TargetUnion::clone" -> "function_outputs::TargetUnion" [label="uses", color="#7f7f7f", style="dashed"] [constraint=false]; // "uses" edge
    "function_outputs::TypeAlias" -> "function_outputs::TargetStruct" [label="uses", color="#7f7f7f", style="dashed"] [constraint=false]; // "uses" edge
    "function_outputs::function" -> "function_outputs::TargetStruct" [label="uses", color="#7f7f7f", style="dashed"] [constraint=false]; // "uses" edge
    "function_outputs::function_output" -> "bool" [label="uses", color="#7f7f7f", style="dashed"] [constraint=false]; // "uses" edge
    "function_outputs::function_output" -> "char" [label="uses", color="#7f7f7f", style="dashed"] [constraint=false]; // "uses" edge
    "function_outputs::function_output" -> "f64" [label="uses", color="#7f7f7f", style="dashed"] [constraint=false]; // "uses" edge
    "function_outputs::function_output" -> "function_outputs::GenericTargetEnum" [label="uses", color="#7f7f7f", style="dashed"] [constraint=false]; // "uses" edge
    "function_outputs::function_output" -> "function_outputs::GenericTargetStruct" [label="uses", color="#7f7f7f", style="dashed"] [constraint=false]; // "uses" edge
    "function_outputs::function_output" -> "function_outputs::GenericTargetUnion" [label="uses", color="#7f7f7f", style="dashed"] [constraint=false]; // "uses" edge
    "function_outputs::function_output" -> "function_outputs::TargetEnum" [label="uses", color="#7f7f7f", style="dashed"] [constraint=false]; // "uses" edge
    "function_outputs::function_output" -> "function_outputs::TargetStruct" [label="uses", color="#7f7f7f", style="dashed"] [constraint=false]; // "uses" edge
    "function_outputs::function_output" -> "function_outputs::TargetTrait" [label="uses", color="#7f7f7f", style="dashed"] [constraint=false]; // "uses" edge
    "function_outputs::function_output" -> "function_outputs::TargetUnion" [label="uses", color="#7f7f7f", style="dashed"] [constraint=false]; // "uses" edge
    "function_outputs::function_output" -> "isize" [label="uses", color="#7f7f7f", style="dashed"] [constraint=false]; // "uses" edge
    "function_outputs::function_output" -> "str" [label="uses", color="#7f7f7f", style="dashed"] [constraint=false]; // "uses" edge
    "function_outputs::function_output" -> "usize" [label="uses", color="#7f7f7f", style="dashed"] [constraint=false]; // "uses" edge

}
